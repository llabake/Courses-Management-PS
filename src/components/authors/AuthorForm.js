import React from 'react';
import {
  string,
  shape,
  arrayOf,
  bool,
  func,
  object,
} from 'prop-types';
import TextInput from '../common/TextInput';

const AuthorForm = ({
  author,
  onSave,
  onChange,
  saving,
  errors,
}) => (
  <form>
    <h1>Manage Authors</h1>
    <TextInput
        name="firstName"
        label="First Name"
        value={author.firstName}
        onChange={onChange}
        error={errors.firstName}
      />

    <TextInput
      name="lastName"
      label="Last Name"
      value={author.lastName}
      onChange={onChange}
      error={errors.lastName}
      />

    <input
        type="submit"
        disabled={saving}
        value={saving ? 'Saving' : 'Save'}
        className="btn btn-primary"
        onClick={onSave}
      />
  </form>
);

export default AuthorForm;
