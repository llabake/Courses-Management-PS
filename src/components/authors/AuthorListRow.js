/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const AuthorListRow = ({ author, onDelete }) => {
  const deleteAuthor = () => (
    onDelete(author.id)
  );
  return (
    <tr>
      <td><Link to={`/author/${author.id}`}>{author.id}</Link></td>
      <td>
        {author.firstName} 
        {' '}
        {author.lastName}
      </td>
      <td
        className="btn btn-xs btn-danger"
        onClick={deleteAuthor}
      >
        Delete
      </td>
    </tr>
  );
};

AuthorListRow.propTypes = {
  author: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default AuthorListRow;
