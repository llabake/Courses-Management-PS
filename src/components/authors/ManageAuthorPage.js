import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Prompt } from 'react-router-dom';
import toastr from "toastr";
import {
  bool,
  shape,
  arrayOf,
  func,
} from 'prop-types';
import { authorInputValidator } from "../../helpers";
import { loadAuthors, saveAuthor } from "../../actions/authorActions";
import AuthorForm from "./AuthorForm";

export class ManageAuthorPage extends Component {
  state = {
    author: Object.assign({}, this.props.author),
    errors: {},
    saving: false,
    isBlocking: false,
    isValid: false,
  };

  updateAuthorState = (event) => {
    const field = event.target.name;
    const { author } = this.state;
    author[field] = event.target.value;
    return this.setState({ author, isBlocking: true });
  };

  saveAuthor = (event) => {
    event.preventDefault();
    if (!this.validate()) {
      return;
    }
    this.setState({ saving: true });
    this.props.saveAuthor(this.state.author)
      .then(() => this.redirect())
      .catch((error) => {
        toastr.error(error);
        this.setState({ saving: false });
      });
  };

  validate() {
    const { errors, isValid } = authorInputValidator(this.state);
    this.setState({ isValid, errors });
    return isValid;
  }

  redirect() {
    this.setState({
      saving: false,
      isBlocking: false,
    });
    toastr.success("Author saved successfully");
    this.props.history.push('/authors');
  }

  render() {
    const {
      author, errors, isBlocking, saving, 
    } = this.state;
    return (
      <div>
        <Prompt
          when={isBlocking}
          message="Are you sure you want to go to leave the page"
        />
        <AuthorForm
          author={author}
          errors={errors}
          onSave={this.saveAuthor}
          saving={saving}
          onChange={this.updateAuthorState}
        />
      </div>
    );
  }
}

export const mapStateToProps = (state, { match: { params } }) => {
  const authorId = params.id;
  let author = {
    id: '', firstName: '', lastName: '',
  };

  if (authorId && state.authors.length > 0) {
    author = state.authors.find(authorObj => authorObj.id === authorId);
  }

  return {
    author,
    authors: state.authors,
  };
};

export const mapDispatchToProps = dispatch => ({
  saveAuthor: author => dispatch(saveAuthor(author)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ManageAuthorPage);
