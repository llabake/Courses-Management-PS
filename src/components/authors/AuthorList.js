import React from 'react';
import PropTypes from 'prop-types';
import AuthorListRow from "./AuthorListRow";

const AuthorList = ({ authors, onDelete }) => (
  <table className="table">
    <thead>
      <tr>
        <th>ID</th>
        <th>Name</th>
        <th />
      </tr>
    </thead>
    <tbody>
      {authors.map(author => (
        <AuthorListRow
          key={author.id}
          author={author}
          onDelete={onDelete} />
      ))}
    </tbody>
  </table>
);

AuthorList.propTypes = {
  authors: PropTypes.array.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default AuthorList;
