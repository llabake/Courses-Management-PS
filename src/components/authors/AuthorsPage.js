import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from "toastr";
import NoCourses from "../common/NoCourses";
import AuthorList from "./AuthorList";
import { loadAuthors, deleteAuthor } from "../../actions/authorActions";

class AuthorsPage extends Component {
  static authorRow(author, index) {
    return <div key={index}>{author.id}</div>;
  }

  state = {
    author: {},
  };

  redirectToAddAuthorPage = () => {
    const { history } = this.props;
    history.push('/author');
  };

  authorCourseExist = (courses, authorId) => {
    return courses.some(course => (course.authorId === authorId));
  };

  deleteAuthor = (authorId) => {
    const { history, courses } = this.props;
    if (!this.authorCourseExist(courses, authorId)) {
      this.props.deleteAuthor(authorId).then(() => {
        toastr.success('Author deleted successfully');
        history.push('/authors');
      });
    } else {
      toastr.warning(`This author can not be deleted as they have a course`);
    }
  };

  render() {
    const { authors, loading } = this.props;
    return (
      <div>
        <h1>Authors</h1>
        <input
          type="submit"
          value="Add Author"
          className="btn btn-primary"
          onClick={this.redirectToAddAuthorPage}
        />
        { !loading && !authors.length
          ? (
            <NoCourses
              text="There are no Authors now. Click the Add Author button to add an author" />
          )
          : (
            <AuthorList
            onDelete={this.deleteAuthor}
            authors={authors}
          />
          )
        }
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  courses: state.courses,
  loading: state.ajaxCallsInProgress > 0,
  authors: state.authors,
});

export const mapDispatchToProps = dispatch => ({
  deleteAuthor: author => dispatch(deleteAuthor(author)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthorsPage);
