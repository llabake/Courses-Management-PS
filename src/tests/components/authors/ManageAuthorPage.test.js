import React from 'react';
import { shallow } from 'enzyme';
import { ManageAuthorPage, mapStateToProps, mapDispatchToProps } from "../../../components/authors/ManageAuthorPage";
import courseMock from "../../mocks/courseMock";
import authorMock from "../../mocks/authorMock";

describe('<ManageAuthorPage', () => {
  it('renders correctly', () => {
    const state = {
      author: {},
      errors: {},
      saving: false,
      isBlocking: false,
      isValid: false,
    };
    const wrapper = shallow(<ManageAuthorPage {...state} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should save an author', () => {
    const props = {
      author: {},
      authors: authorMock.authors,
      match: {
        params: {
          id: courseMock.courses[0].id,
        },
      },
      saveAuthor: jest.fn(() => Promise.resolve()),
      updateAuthorState: jest.fn(() => Promise.resolve()),
      history: { push: jest.fn() },
    };
    const wrapper = shallow(<ManageAuthorPage {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'saveAuthor');
    wrapper.setState({
      author: {
        firstName: 'my author name',
        lastName: 'my author last name',
      },
    });
    wrapper.instance().saveAuthor({ preventDefault: jest.fn() });
    expect(spy).toHaveBeenCalled();
  });
  it('renders a Prompt child component', () => {
    const state = {
      author: {},
      errors: {},
      saving: false,
      isBlocking: false,
      isValid: false,
    };

    const props = {
      saveAuthor: jest.fn(),
      updateAuthorState: jest.fn(),
      history: { push: jest.fn() },
    };
    const wrapper = shallow(<ManageAuthorPage {...state} {...props} />);
    expect(wrapper.find('Prompt').length).toBe(1);
  });
  it('should update an author', () => {
    const props = {
      course: courseMock.courses[0],
      authors: authorMock.authors,
      author: authorMock.authors[0],
      match: {
        params: {
          id: authorMock.authors[0].id,
        },
      },
      saveAuthor: jest.fn(() => Promise.resolve()),
      updateAuthorState: jest.fn(() => Promise.resolve()),
      history: { push: jest.fn() },
    };
    const wrapper = shallow(<ManageAuthorPage {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'updateAuthorState');
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'firstName',
        value: 'well well',
      },
    };
    // wrapper.setState({
    //   course: {
    //     title: 'my course title',
    //     length: '07:10',
    //     authorId: 'my author',
    //     category: 'my category',
    //   }
    // });
    wrapper.instance().updateAuthorState(event);
    expect(spy).toHaveBeenCalled();
  });
  it('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('saveAuthor');
    const { saveAuthor } = mapDispatchToProps(dispatch);
    saveAuthor(authorMock.createdAuthorWithId);
    expect(dispatch).toHaveBeenCalled();
  });
});
