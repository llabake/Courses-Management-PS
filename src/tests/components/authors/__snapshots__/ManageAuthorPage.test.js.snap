// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageAuthorPage renders correctly 1`] = `
<div>
  <Prompt
    message="Are you sure you want to go to leave the page"
    when={false}
  />
  <AuthorForm
    author={Object {}}
    errors={Object {}}
    onChange={[Function]}
    onSave={[Function]}
    saving={false}
  />
</div>
`;
